File: rotate.ts
inlets = 3;
outlets = 2;

var seqReal: number[] = [];
var seqImg: number[] = [];
var angle = 0;

function msg_float(v: number)
{
	switch (inlet) {
        case 2:
            angle = v;
            break;
	}
	bang();
}

function list()
{
	switch (inlet) {
		case 0:
            seqReal = arrayfromargs(arguments);
            bang();
			break;
        case 1:
            seqImg = arrayfromargs(arguments);
            break;
	}
}

var cartopol = function (x: number, y: number) {
    return ({
        t: Math.atan2(y, x),
        r: Math.sqrt((x * x) + (y * y))
    });
};

var poltocar = function (t: number, r: number) {  
    return ({
        x: r * Math.cos(t),
        y: r * Math.sin(t)
    });
};

function bang()
{
    var len = seqReal.length;
    var outSeqReal = [];
    var outSeqImg = [];

    for(let i = 0; i < len; i++){
        var pol = cartopol(seqReal[i], seqImg[i]);
        var outCar = poltocar(pol.t - angle, pol.r);
        outSeqReal.push(outCar.x);
        outSeqImg.push(outCar.y);
    }    

    outlet(1, outSeqImg);
	outlet(0, outSeqReal);
}

let module = {};
export = {};

File: rotate.java
import com.cycling74.max.*;

public class rotate extends MaxObject {

    float[] seqReal, seqImg;
    float angle;

    rotate() {
        declareInlets(new int[] { DataTypes.LIST, DataTypes.LIST, DataTypes.FLOAT });
        declareOutlets(new int[] { DataTypes.LIST, DataTypes.LIST });
    }

    protected void inlet(float v) {
        final int idx = getInlet();
        switch (idx) {
            case 2:
                angle = v;
                bang();
                break;
        }
    }

    protected void list(Atom[] a) {
        final int idx = getInlet();
        switch (idx) {
            case 0:
                seqReal = Atom.toFloat(a);
                bang();
                break;
            case 1:
                seqImg = Atom.toFloat(a);
                break;
        }
    }

    private class CarToPol {
        float t, r;

        CarToPol(float x, float y) {
            t = (float) Math.atan2(y, x);
            r = (float) Math.sqrt((x * x) + (y * y));
        }
    }

    private class PolToCar {
        float x, y;

        PolToCar(float t, float r) {
            x = (float) (r * Math.cos(t));
            y = (float) (r * Math.sin(t));
        }
    }

    protected void bang() {
        int len = seqReal.length;
        float[] outSeqReal = new float[len];
        float[] outSeqImg = new float[len];

        for (int i = 0; i < len; i++) {
            CarToPol pol = new CarToPol(seqReal[i], seqImg[i]);
            PolToCar car = new PolToCar(pol.t - angle, pol.r);
            outSeqReal[i] = car.x;
            outSeqImg[i] = car.y;
        }

        outlet(1, outSeqImg);
        outlet(0, outSeqReal);
    }

}
